src/server/session.ts

```diff
@@ -1973,7 +1973,7 @@ export class Session<TMessage = string> implements EventSender {
 
     private getReferences(args: protocol.FileLocationRequestArgs, simplifiedResult: boolean): protocol.ReferencesResponseBody | readonly ReferencedSymbol[] {
         const file = toNormalizedPath(args.file);
-        const projects = this.getProjects(args);
+        const projects = this.getProjects(args, /*getScriptInfoEnsuringProjectsUptoDate*/ true);

```

Comment:
I think instead you want to guard against the project being upto date at the usage site. The projects can become dirty at any moement (eg when new project is created because of which open file goes from inferred project to that configured/inferred project - as part of finding original location of a file). This will also avoid having to update all projects and confine the updates to affected oprojects.

You want to use `updateIfDirty`

Reply:
I thought about that, but right here in this call to `this.getProjects(args)`, we already might be using outdated information, because we use the argument's file's `scriptInfo.containingProjects` to decide what projects we pass down to `getReferencesWorker`, so I thought we needed to use the updated information here. Or is it ok to use outdated information here when picking the projects to pass to `getReferencesWorker`, because we will somehow make up for it inside `getReferencesWorker`?

Just to make sure I understand, are you saying that projects can become dirty during the find all references work?
And by "project being up to date at the usage site", do you mean when we enqueue a project+location, when we dequeue a project to use it, or something else?

Reply:
> I thought about that, but right here in this call to this.getProjects(args), we already might be using outdated information, because we use the argument's file's scriptInfo.containingProjects to decide what projects we pass down to getReferencesWorker, so I thought we needed to use the updated information here. Or is it ok to use outdated information here when picking the projects to pass to getReferencesWorker, because we will somehow make up for it inside getReferencesWorker?

Yes its ok to not update all the projects as thats lot of work. You want to start with whatever projects are there and then go from there.

>do you mean when we enqueue a project+location, when we dequeue a project to use it, or something else?

Yes at the time of dequeue you want to check if the project can really be used or not. (based on updateIfDirty and file being still part of that project)